Implement Send Invoice flow from Global Create

VickyStash opened the pull request:
### Details
<!-- Explanation of the change or anything fishy that is going on -->
Implementation of invoice creation flow from Global Create

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/40012
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

**Note (!):** Use staging server by turning it on in `Settings` -> `Troubleshoot` -> `Use Staging Server`
**Note (!):** the backend has a know [issue](https://github.com/Expensify/App/issues/41068) related to the invoice currency. If you want your test results to not be affected by that issue, set the sender workspace and the invoice currencies to be USD.

**Flow 1**

1. Sign in to the app with a workspace admin account
2.  Open Global Create and verify `Send Invoice` appears in Global Create
3.  Click Send invoice in Global Create:
3.1 Verify “Distance” and “Scan" do not appear in the big number pad (BNP) screen.
3.2 Verify the Currency can be modified
4. Add an amount  -> tap Next
5. Select the participant, and make sure the participant's step works as expected. Press Next.
6. Verify all of the information on the confirmation screen looks as expected.
6.1 The sender Workspace is displayed as expected.
6.2 The receiver is displayed as expected.
7. Tap on the `Send from` workspace, you should be navigated to the Send from the screen.
8. Select another workspace, and make sure it works as expected.
9.  tap Send $XX.XX invoice, verify all of the necessary data is created in Onyx.

**Flow 2**
1. Repeat Flow 1, but select the same sender and receiver.
2. Make sure you are navigated to the same room after invoice confirmation.

**Flow 3**
1. Sign in to the app with an account without workspaces
3. Confirm you have no `Send Invoice` option in Global Create

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as in the **Tests** section

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
4. Upload an image via copy paste
11. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

Same as in the **Tests** section

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/23176449/6b690c1a-9bdf-4f0b-8ef8-746b09e688c3


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/23176449/7e2e98ba-64b4-47a9-a6cf-008097071c2e


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/23176449/9aed0d1c-527b-4d61-a6b7-c9506621f751



</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/23176449/9f0479cb-cc97-450c-9752-55e755d730fb


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/23176449/dabd2cec-add8-478b-816a-11079abd3e62



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->




https://github.com/Expensify/App/assets/23176449/adb5e566-a501-4531-9ae8-fd5dc17b0140



</details>

VickyStash committed:
Partly update start, participants and confirmation steps to support invoices

VickyStash committed:
Implement send from (select sender) screen

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/pages/iou/request/IOURequestStartPage.js
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js
#	src/pages/iou/request/step/withWritableReportOrNotFound.tsx

VickyStash committed:
Implement send invoice functionality

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js
#	src/pages/iou/request/step/withFullTransactionOrNotFound.tsx
#	src/pages/iou/request/step/withWritableReportOrNotFound.tsx

VickyStash committed:
Fix ts issues after merging main

VickyStash commented on src/languages/es.ts:
Need the confirmation regarding this translations

VickyStash committed:
Code improvements

VickyStash committed:
IOU file updates and lint fixes

VickyStash committed:
Implement buildOptimisticInviteReportAction function

VickyStash committed:
Fix navigation

VickyStash committed:
Fix invoice creation from global in invoice chat room already exists

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/libs/API/parameters/index.ts
#	src/libs/API/types.ts
#	src/libs/Navigation/types.ts
#	src/libs/ReportUtils.ts
#	src/pages/iou/request/IOURequestStartPage.tsx
#	src/pages/iou/request/step/withFullTransactionOrNotFound.tsx

VickyStash committed:
Minor lint fix

VickyStash committed:
Enable add receipt file feature, lint fixes

VickyStash committed:
Minor UI fixes

VickyStash committed:
Fix translation

VickyStash committed:
Fix report preview message

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js

VickyStash committed:
Minor code improvements

VickyStash committed:
Fix canSendInvoice check

VickyStash committed:
Minor code improvements

VickyStash committed:
Put iouCreatedAction to optimistic data

rezkiy37 commented on src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:
```diff
@@ -240,6 +247,16 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
+    }, [allPolicies, pickedParticipants]);
+
+    const canUpdateSenderWorkspace = useMemo(
+        () => PolicyUtils.getActiveAdminWorkspaces(allPolicies).length > 0 && !!transaction?.isFromGlobalCreate,
```
```suggestion
        () => PolicyUtils.canSendInvoice(allPolicies) > 0 && !!transaction?.isFromGlobalCreate,
```

rezkiy37 commented on src/libs/actions/IOU.ts:
```diff
@@ -805,6 +806,306 @@ function buildOnyxDataForMoneyRequest(
 
+/** Builds the Onyx data for an invoice */
+function buildOnyxDataForInvoice(
+    chatReport: OnyxEntry<OnyxTypes.Report>,
+    iouReport: OnyxTypes.Report,
+    transaction: OnyxTypes.Transaction,
+    chatCreatedAction: OptimisticCreatedReportAction,
+    iouCreatedAction: OptimisticCreatedReportAction,
+    iouAction: OptimisticIOUReportAction,
+    reportPreviewAction: ReportAction,
+    optimisticPolicyRecentlyUsedCategories: string[],
+    optimisticPolicyRecentlyUsedTags: OnyxTypes.RecentlyUsedTags,
+    isNewChatReport: boolean,
+    transactionThreadReport: OptimisticChatReport,
+    transactionThreadCreatedReportAction: OptimisticCreatedReportAction,
+    inviteReportAction?: OptimisticInviteReportAction,
+    policy?: OnyxEntry<OnyxTypes.Policy>,
+    policyTagList?: OnyxEntry<OnyxTypes.PolicyTagList>,
+    policyCategories?: OnyxEntry<OnyxTypes.PolicyCategories>,
```
Can we use an object instead of these arguments?

VickyStash replied:
It's the same way in other places in this file, so I decided to keep it consistent. Let's see how further review will go, I'll do quick update if it's necessary

rezkiy37 commented on src/libs/actions/IOU.ts:
```diff
@@ -805,6 +806,306 @@ function buildOnyxDataForMoneyRequest(
+        });
+    }
+
+    optimisticData.push(
```
If something we can add without any conditions, let's set it by default to avoid an extra `push`.

rezkiy37 commented on src/libs/actions/IOU.ts:
```diff
@@ -805,6 +806,306 @@ function buildOnyxDataForMoneyRequest(
+        });
+    }
+
+    successData.push(
```
If something we can add without any conditions, let's set it by default to avoid an extra `push`.

rezkiy37 commented on src/libs/actions/IOU.ts:
```diff
@@ -1105,6 +1406,125 @@ function buildOnyxDataForTrackExpense(
 
+/** Gathers all the data needed to create an invoice. */
+function getSendInvoiceInformation(
+    transaction: OnyxEntry<OnyxTypes.Transaction>,
+    invoiceChatReport?: OnyxEntry<OnyxTypes.Report>,
+    receipt?: Receipt,
+    policy?: OnyxEntry<OnyxTypes.Policy>,
+    policyTagList?: OnyxEntry<OnyxTypes.PolicyTagList>,
+    policyCategories?: OnyxEntry<OnyxTypes.PolicyCategories>,
```
An object instead of many arguments?

rezkiy37 commented on src/libs/actions/IOU.ts:
```diff
@@ -2363,6 +2783,64 @@ function requestMoney(
 }
 
+function sendInvoice(
+    currentUserAccountID: number,
+    transaction: OnyxEntry<OnyxTypes.Transaction>,
+    invoiceChatReport?: OnyxEntry<OnyxTypes.Report>,
+    receiptFile?: Receipt,
+    policy?: OnyxEntry<OnyxTypes.Policy>,
+    policyTagList?: OnyxEntry<OnyxTypes.PolicyTagList>,
+    policyCategories?: OnyxEntry<OnyxTypes.PolicyCategories>,
```
An object instead of many arguments?

VickyStash replied:
It's the same way in other places in this file, so I decided to keep it consistent. Let's see how further review will go, I'll do quick update if it's necessary

rezkiy37 reviewed (changes requested):
Left comments.

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx
#	src/libs/PolicyUtils.ts
#	src/libs/actions/IOU.ts
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js
#	src/pages/iou/request/step/IOURequestStepConfirmation.tsx

VickyStash committed:
Minor improvements

VickyStash committed:
Update optimistic invoice room creation to include current user as a member

VickyStash committed:
Add INVOICE_RECEIVER_TYPE to consts

VickyStash committed:
Code improvements

VickyStash committed:
Add optimistic personal details for the receiver; update params

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/languages/en.ts
#	src/languages/es.ts
#	src/libs/API/parameters/index.ts
#	src/libs/ReportUtils.ts
#	src/pages/iou/request/IOURequestStartPage.tsx

VickyStash committed:
Add SendInvoiceInformation type

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx
#	src/libs/ReportUtils.ts
#	src/libs/actions/IOU.ts

VickyStash committed:
Fix to follow typing rules in ROUTES

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.tsx

VickyStash committed:
Update param invoiceRoomID -> invoiceRoomReportID

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js

VickyStash committed:
Updates to follow main branch changes

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/SCREENS.ts
#	src/libs/Navigation/AppNavigator/ModalStackNavigators/index.tsx
#	src/libs/Navigation/types.ts

VickyStash committed:
Lint fix

VickyStash commented:
**Updates:**
- Updated param invoiceRoomID -> invoiceRoomReportID
- Updates to follow main branch changes (MoneyTemporaryForRefactorRequestConfirmationList.tsx got removed)
- Tested API updates

**TODOs:**
- Waiting for API fixes to be able to progress PR

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/components/MoneyRequestConfirmationList.tsx
#	src/libs/PolicyUtils.ts
#	src/libs/ReportUtils.ts
#	src/libs/actions/IOU.ts
#	src/pages/iou/request/IOURequestStartPage.tsx
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js
#	src/pages/iou/request/step/IOURequestStepParticipants.tsx

VickyStash committed:
Update sender logic to work with updates in main

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/components/MoneyRequestConfirmationList.tsx

VickyStash committed:
Lint fix

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/CONST.ts
#	src/components/MoneyRequestConfirmationList.tsx
#	src/components/ReportWelcomeText.tsx
#	src/libs/IOUUtils.ts
#	src/libs/Navigation/types.ts
#	src/libs/ReportUtils.ts
#	src/pages/iou/request/IOURequestStartPage.tsx
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js
#	src/pages/iou/request/step/IOURequestStepConfirmation.tsx
#	src/pages/iou/request/step/IOURequestStepParticipants.tsx

VickyStash committed:
TS fixes

VickyStash committed:
Lint fixes

VickyStash commented:
This PR is still waiting for some API fixes, but due to the short timeline, we need to start the review process the sooner the better!
cc @cristipaval

shawnborton commented:
For the big number pad screen, we decided that there should be no tab control at the top:
![CleanShot 2024-04-24 at 11 01 22@2x](https://github.com/Expensify/App/assets/2319350/a7f0c5a5-a9fb-4b1e-933d-36308a5a0313)

shawnborton commented:
Also just wanted to double check with @JmillsExpensify @dannymcclain @davidcardoza - are we going to add the Next steps right into the preview card? Looks like this PR has them:
![CleanShot 2024-04-24 at 11 03 22@2x](https://github.com/Expensify/App/assets/2319350/f7a2497f-9f5f-470d-8206-4bbb3536a47b)

I thought we might have decided to punt that?

VickyStash commented:
@shawnborton Unfortunately, there are some API known issues that block me from doing the correct  screenshots/videos without the offline mode. But it should be resolved tomorrow morning, and I'll update the recordings as soon as it will be done

shawnborton commented:
Sounds good!

dannymcclain commented:
>Also just wanted to double check with @JmillsExpensify @dannymcclain @davidcardoza - are we going to add the Next steps right into the preview card?
>I thought we might have decided to punt that?

Yes I believe we did decide to not tackle that right now.

VickyStash commented:
@shawnborton @dannymcclain 

> Yes I believe we did decide to not tackle that right now.

Just to confirm, we shouldn't show `Awaiting payment by xxx` for now, right?

^^ @rezkiy37

shawnborton commented:
Let's get confirmation from @JmillsExpensify on that first. If we implement it for invoices, we should be implementing it for all preview cards as well.

davidcardoza commented:
Brought the topic up in the VIP room - https://expensify.slack.com/archives/CSL3XBCCR/p1713994340869369

VickyStash committed:
Hide tabs on the amount step for invoices

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/libs/PolicyUtils.ts

VickyStash committed:
Fixes after merging main

shawnborton commented on assets/images/invoice-generic.svg:
```diff
@@ -0,0 +1,4 @@
+<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
+<path d="M6.75 7C6.33579 7 6 7.33579 6 7.75C6 8.16421 6.33579 8.5 6.75 8.5H13.25C13.6642 8.5 14 8.16421 14 7.75C14 7.33579 13.6642 7 13.25 7H6.75Z" fill="#03D47C"/>
+<path fill-rule="evenodd" clip-rule="evenodd" d="M3.8254 1.70739L4.75 2.49991L6.1746 1.27882C6.36185 1.11832 6.63815 1.11832 6.8254 1.27882L8.25 2.49991L9.6746 1.27882C9.86185 1.11832 10.1382 1.11832 10.3254 1.27882L11.75 2.49991L13.1746 1.27882C13.3618 1.11832 13.6382 1.11832 13.8254 1.27882L15.25 2.49991L16.1746 1.70739C16.4989 1.42939 17 1.65984 17 2.08702V7.1875L17.6306 7.69986C17.8643 7.8897 ...
+</svg>
```
We usually supply them at pure black (`#000000`) - maybe we just need to edit the fill value here as a quick fix? Otherwise I am happy to provide a new asset.

shawnborton replied:
Here is an update, exported the same exact way I've exported many other icons in the repo. 
[invoice-generic.svg.zip](https://github.com/Expensify/App/files/15124398/invoice-generic.svg.zip)

VickyStash replied:
Updated to use provided icon - https://github.com/Expensify/App/pull/40015/commits/e22c3eb565a51ad21ddc94eaaba237cce3358a2d

shawnborton commented on src/pages/iou/request/step/IOURequestStepSendFrom.tsx:
```diff
@@ -0,0 +1,123 @@
+        navigateBack();
+    };
+
+    const renderCheckbox = useCallback(
+        (item: ListItem) => (
+            <View style={[styles.roundCheckmarkWrapper, styles.mh2]}>
+                {item.isSelected && (
+                    <Icon
+                        src={Expensicons.Checkmark}
+                        fill={theme.success}
+                    />
+                )}
+            </View>
+        ),
+        [styles.roundCheckmarkWrapper, styles.mh2, theme.success],
+    );
```
Can you share a screenshot of what you are referring to?

I think I am following you though - I suppose we could make the selection look more like this screen?

![CleanShot 2024-04-25 at 21 29 05@2x](https://github.com/Expensify/App/assets/2319350/37bdc39b-2555-41f2-8684-795da7537b06)

shawnborton replied:
Got it I don't feel strongly, I would be totally down to use the other list style I referenced above.

VickyStash replied:
Updated to use default SectionList checkmark - https://github.com/Expensify/App/pull/40015/commits/f3d02339a9c32f2d893660effa4753ae3dd4d861
![image](https://github.com/Expensify/App/assets/23176449/2d5ac2b1-6c40-4432-964b-7e880be79a45)

dannymcclain replied:
Good shout. I think this should just use our regular select list styles—not multi-select. Examples below.

![CleanShot 2024-04-26 at 08 33 36@2x](https://github.com/Expensify/App/assets/7516624/7a8f2844-d02d-40a7-ae6e-d4d5c99b629b)

![CleanShot 2024-04-26 at 08 35 07@2x](https://github.com/Expensify/App/assets/7516624/3ccf31ac-5db1-468a-95af-ef8d1d39ca4b)

![CleanShot 2024-04-26 at 08 35 57@2x](https://github.com/Expensify/App/assets/7516624/fe7f9819-74cd-4a0d-86d9-ff7fd25020f6)

VickyStash replied:
@dannymcclain Updated a little bit more to follow your screenshots 🙂 
![image](https://github.com/Expensify/App/assets/23176449/8e3124e7-3da9-4d32-b0eb-f7948e53cd04)

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.tsx
#	src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js

VickyStash committed:
Update invoice-generic.svg icon

VickyStash committed:
Use default SectionList checkmark on the SendFrom screen

VickyStash committed:
Remove unused style

VickyStash committed:
Don't show three dots on the confirmation screen

VickyStash commented:
@shubham1206agra I've applied your feedback, please take another look!
I've also updated Notes in the PR **Tests** section, so please take a look at them before testing.
Updated recordings in online mode have also been attached.

VickyStash committed:
Update Send From list display

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/libs/API/parameters/index.ts

VickyStash committed:
Lint fix

cristipaval commented:
@shawnborton @dannymcclain is this PR approved from the design standpoint?

cristipaval commented:
> > the display of Invoice room and related report actions is implemented in this PR #40303. So until the PR is not merged, you need to pull it manually to this for testing.
> 
> @VickyStash Do I need to do anything extra or changes are already present here?

@shubham1206agra I held this one on https://github.com/Expensify/App/pull/40303

VickyStash commented:
> @VickyStash Do I need to do anything extra or changes are already present here?

@shubham1206agra You don't need to make any additional changes, maybe just minor conflict resolution can be required

cristipaval commented:
Wooot! 🎉  [The other PR](https://github.com/Expensify/App/pull/40303) is merged. Let's pull `main` into this one and test the ent-to-end Send Invoice flow

VickyStash committed:
Merge branch 'main' into feature/global-send-invoice

# Conflicts:
#	src/CONST.ts
#	src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.tsx

VickyStash committed:
Remove duplicates

VickyStash commented:
@shubham1206agra I've updated the PR, please take a look 👀

VickyStash committed:
Fix invoice report name display

dubielzyk-expensify commented:
Looks good at my first glance, but I know @shawnborton is closer to this, so I'll let him chime in

VickyStash commented:
@shubham1206agra 

> And why did we allowed addition of receipts in this?

Could you please clarify where were you able to ass a receipt?

VickyStash committed:
Hide payment button for invoices

VickyStash committed:
Disable invoice editing for the invoice receiver

VickyStash committed:
Disable invoice receipt update

VickyStash commented:
> The subtitle seems to be broken.

@shubham1206agra Could you please let me know what is the expected subtitle?

VickyStash commented:
> See 2 `Shubham:` written

Sorry, I don't get it, could you please clarify

cristipaval commented:
> @VickyStash Erroneous total and GBR showing while editing amount in invoice room. cc @cristipaval
@shubham1206agra the wrong amount after transaction edit is due to a backend issue that we fixed but it is not deployed yet. We are aware of it so please do not block this PR on this issue 🙏

VickyStash commented:
> No GBR present here.

@shubham1206agra GBR and Pay button are out of scope, I've hidden Pay button as well in this commit https://github.com/Expensify/App/pull/40015/commits/c36bd3bec9a56c42b628ed7d188cbfd85061c4f6

VickyStash commented:
> <img alt="Screenshot 2024-04-29 at 11 59 00 AM" width="1512" src="https://private-user-images.githubusercontent.com/58412969/326345959-2ae8525c-a5f0-47b9-b0e5-fbc81a0156f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzODA0MDgsIm5iZiI6MTcxNDM4MDEwOCwicGF0aCI6Ii81ODQxMjk2OS8zMjYzNDU5NTktMmFlODUyNWMtYTVmMC00N2I5LWIwZTUtZmJjODFhMDE1NmYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI5VDA4NDE0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiOTkzZDVjNWVhODZmZmI2NDZmNDIzMTg1N2Y0NmFlMmFiZGI2ZGFjNTM2NGIxYTVlMGVhZTgzNGYzYmEyNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DwZACIQDZM1t-SRi-6HrkioyWLJyrKNMjcJ4sJw--js"> <img alt="Screenshot 2024-04-29 at 11 59 23 AM" width="1512" src="https://private-user-images.githubusercontent.com/58412969/326345969-2c9816de-a216-4d30-9f7b-4bc7c160d7fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzODA0MDgsIm5iZiI6MTcxNDM4MDEwOCwicGF0aCI6Ii81ODQxMjk2OS8zMjYzNDU5NjktMmM5ODE2ZGUtYTIxNi00ZDMwLTlmN2ItNGJjN2MxNjBkN2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI5VDA4NDE0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzODUyNzkxYWY3MDY1YTA4YjkyMWM4ZDVjOTFmYTM3YjMwOWYwNjQ2NGJkZDgxMzY4MTBiOWVjZjJkMmIzMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Lp321qMExk9fDOGYydzmb-1--vScqE7E_Vnm6h75uQQ">
> After doing `clear cache and restart`, the App crashes on opening invoices. (I have added a receipt in one invoice).

@shubham1206agra I've made some investigations and it looks like it can be related to another issue. There is an open [issue](https://github.com/Expensify/App/issues/41130) with the app crashing with another flow, but the same error. There is a [proposal](https://github.com/Expensify/App/issues/41130#issuecomment-2080881543) that describes the reason of the problem and the solution, so I'm not sure if I should resolve it in the current PR

VickyStash committed:
Fix LHN option display

VickyStash committed:
Remove quick action update during invoice creation

VickyStash commented:
> @VickyStash One minor thing I noticed is that QAB gets updated on Send Invoice. This shouldn't happen.

@shubham1206agra Thank you, fixed - https://github.com/Expensify/App/pull/40015/commits/43710829df28df950ce6b2364ccfade09f4831b3

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

puneetlath commented:
Yes, that's right. Just like workspace chats we'd have the big sender avatar of the sender's workspace with the small payer user avatar.

VickyStash commented:
@cristipaval Can we please create a follow-up issue to update the invoice room avatars display?
cc @waterim 

> Similar comment here as the other PR, I think we might not be using the correct room avatars here? cc @puneetlath
> 
> I think the idea was to do this for Workspace to Workspace: ![image](https://private-user-images.githubusercontent.com/2319350/326518246-84439975-95d5-480a-932b-49cdcc402547.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0MTIzMDYsIm5iZiI6MTcxNDQxMjAwNiwicGF0aCI6Ii8yMzE5MzUwLzMyNjUxODI0Ni04NDQzOTk3NS05NWQ1LTQ4MGEtOTMyYi00OWNkY2M0MDI1NDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjlUMTczMzI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmNkZGI3YzdiYjRhNDA5NjRhMTcyOTE4MDlhYmU4MThhOTc4M2M0MzU0NzZiNGUwZjcxYzVlMjNlZTM4M2YyMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gO6LW27tlBd2YWUUZnmty5bL2dhy26_ejG1eZB6btt0)
> 
> And to do this for Workspace to Individual: ![image](https://private-user-images.githubusercontent.com/2319350/326518293-c97da196-8c5f-427d-a4c1-f2c6f9b4d1d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0MTIzMDYsIm5iZiI6MTcxNDQxMjAwNiwicGF0aCI6Ii8yMzE5MzUwLzMyNjUxODI5My1jOTdkYTE5Ni04YzVmLTQyN2QtYTRjMS1mMmM2ZjliNGQxZDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjlUMTczMzI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODE5M2FjNjU3MWExMTU2NWMyNzZhZmQ5ODMxZWJiZjEwMGMyMjFhOGI2OTNmMDZiMjk1ZDk5NjQwZjdmMjhmNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4RzDA4ClvA-G8jYRfsTQ_HekAObFbn78LuOXPorTlt8)
> 
> Does that sound right Puneet & @davidcardoza ?

danielrvidal commented:
>Can we please create a follow-up issue to update the invoice room avatars display?
I created the issue here: https://github.com/Expensify/App/issues/41261

Let me know if I need to do anything else.

OSBotify commented:
🚀 [Cherry-picked](https://github.com/Expensify/App/actions/runs/8886369540) to staging by https://github.com/francoisl in version: 1.4.67-7 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

@Expensify/applauseleads please QA this PR and check it off on the [deploy checklist](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) if it passes.

OSBotify commented:
🚀 [Cherry-picked](https://github.com/Expensify/App/actions/runs/8886369540) to staging by https://github.com/francoisl in version: 1.4.67-7 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

@Expensify/applauseleads please QA this PR and check it off on the [deploy checklist](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) if it passes.

MonilBhavsar commented:
There's a bug noted mostly coming from this PR https://github.com/Expensify/App/issues/41287. If anyone can please take a look

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8904495825) to production by https://github.com/francoisl in version: 1.4.68-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

roryabraham commented on src/libs/API/parameters/SendInvoiceParams.ts:
```diff
@@ -0,0 +1,19 @@
+type SendInvoiceParams = {
+    senderWorkspaceID: string;
+    accountID: number;
+    receiverEmail?: string;
```
Nice little pro tip: _either_ `receiverEmail` _or_ `receiverInvoiceRoomID` is required. So I see why you made them both optional, but it later led to a problem [here](https://github.com/Expensify/App/issues/43060). So what you can do is use [RequireAtLeastOne](https://github.com/sindresorhus/type-fest/blob/main/source/require-at-least-one.d.ts) from type-fest.

Pujan92 commented:
We missed considering the enabled features of the collect workspace to show in the send invoice confirmation screen which raised an issue [here](https://github.com/Expensify/App/issues/41281).

rojiphil commented on src/libs/actions/Policy.ts:
```diff
@@ -277,6 +277,16 @@ function getPolicy(policyID: string | undefined): Policy | EmptyObject {
+ */
+function getPrimaryPolicy(activePolicyID?: string): Policy | undefined {
+    const activeAdminWorkspaces = PolicyUtils.getActiveAdminWorkspaces(allPolicies);
+    const primaryPolicy: Policy | null | undefined = allPolicies?.[activePolicyID ?? ''];
```
As we did not exclude personal policies while fetching primary policy, this resulted in #46705

mollfpr commented on src/libs/actions/IOU.ts:
```diff
@@ -780,6 +793,301 @@ function buildOnyxDataForMoneyRequest(
+            key: `${ONYXKEYS.COLLECTION.TRANSACTION}${transaction.transactionID}`,
+            value: {
+                errors: ErrorUtils.getMicroSecondOnyxError('iou.error.genericCreateInvoiceFailureMessage'),
+                pendingAction: null,
```
✋ Coming from https://github.com/Expensify/App/issues/54035

We already have a logic to delete the transaction if the transaction pending action is ADD when closing the error.

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/components/ReportActionItem/MoneyRequestView.tsx#L458-L472

So, clear the `pendingAction` makes the report not deleted. We only need to clear the `pendingAction` on success.

rojiphil commented on src/libs/actions/IOU.ts:
```diff
@@ -2881,6 +3326,52 @@ function requestMoney(
+    API.write(WRITE_COMMANDS.SEND_INVOICE, parameters, onyxData);
+
+    Navigation.dismissModal(invoiceRoomReportID);
+    Report.notifyNewAction(invoiceRoomReportID, receiver.accountID);
```
Sending a new invoice did not scroll down the view as we missed passing the current account id instead of the receiver's account id. We fixed this in #63089
